WorkloadManagementSystem/Utilities/Pilot3Synchronizer.py

```diff
@@ -0,0 +1,274 @@
+import urllib
+import shutil
+import os
+import glob
+import tarfile
+from git import Repo
+
+from DIRAC                                    import gLogger, S_OK, gConfig, S_ERROR
+from DIRAC.Core.DISET.HTTPDISETConnection     import HTTPDISETConnection
+
+__RCSID__ = '$Id:  $'
+
+
+class Pilot3Synchronizer( object ):
+  '''
+  2 functions are executed:
+  - It updates a JSON file with the values on the CS which can be used by Pilot3 pilots
+  - It updates the pilot 3 files
+
+  This synchronizer can be triggeres at any time via Pilot3Synchronizer().sync().
+  As it is todat, this is triggered every time there is a successful write on the CS.
+  '''
+
+  def __init__( self, paramDict ):
+    ''' c'tor
+
+        Just setting defaults
+    '''
+    self.jsonFile = 'pilot.json'  # default filename of the pilot json file
+    # domain name of the web server used to upload the pilot json file and the pilot scripts
+    self.pilotFileServer = paramDict['pilotFileServer']
+    self.pilotRepo = paramDict['pilotRepo']  # repository of the pilot
+    self.pilotVORepo = paramDict['pilotVORepo']  # repository of the VO that can contain a pilot extension
+    self.pilotLocalRepo = 'pilotLocalRepo'  # local repository to be created
+    self.pilotVOLocalRepo = 'pilotVOLocalRepo'  # local VO repository to be created
+    self.pilotSetup = gConfig.getValue( '/DIRAC/Setup', '' )
+    self.projectDir = paramDict['projectDir']
+    # where the find the pilot scripts in the VO pilot repository
+    self.pilotVOScriptPath = paramDict['pilotVOScriptPath']
+    self.pilotScriptsPath = paramDict['pilotScriptsPath'] # where the find the pilot scripts in the pilot repository
+    self.pilotVersion = ''
+    self.pilotVOVersion =''
+
+  def sync( self ):
+    ''' Main synchronizer method.
+    '''
+    gLogger.notice( '-- Synchronizing the content of the JSON file %s with the content of the CS --' % self.jsonFile )
+
+    result = self._syncFile()
+    if not result['OK']:
+      gLogger.error( "Error uploading the pilot file: %s" % result['Message'] )
+      return result
+
+    gLogger.notice( '-- Synchronizing the pilot scripts %s with the content of the repository --' % self.pilotRepo )
+
+    self._syncScripts()
+
+    return S_OK()
+
+  def _syncFile( self ):
+    ''' Creates the pilot dictionary from the CS, ready for encoding as JSON
+    '''
+    pilotDict = self._getCSDict()
+
+    result = self._upload( pilotDict = pilotDict )
+    if not result['OK']:
+      gLogger.error( "Error uploading the pilot file: %s" %result['Message'] )
+      return result
+    return S_OK()
+
+  def _getCSDict(self):
+    """ Gets minimal info for running a pilot, from the CS
+
+    :returns: pilotDict (containing pilots run info)
+    :rtype: dict
+    """
+
+    pilotDict = { 'Setups' : {}, 'CEs' : {} }
+
+    gLogger.info( '-- Getting the content of the CS --' )
+
+    gLogger.verbose( 'From Operations/[Setup]/Pilot' )
+    setups = gConfig.getSections( '/Operations/' )
```

Comment:
I don't think this works for multi-VO instances.
Why not get this from  "/DIRAC/Setups"?

Reply:
Good point about the multi-VO instances. 

For what regards the setups: they are not taken from "/DIRAC/Setups" because then the "/Operations" section may contain simply "Operations/Defaults"